pkg/gatewayserver/gatewayserver_test.go

```diff
@@ -556,11 +556,18 @@ func TestGatewayServer(t *testing.T) {
 					},
 				} {
 					t.Run(ctc.Name, func(t *testing.T) {
-						ctx, cancel := context.WithDeadline(ctx, time.Now().Add(timeout))
-						upCh := make(chan *ttnpb.GatewayUp)
-						downCh := make(chan *ttnpb.GatewayDown)
-						err := ptc.Link(ctx, t, ctc.ID, ctc.Key, upCh, downCh)
-						cancel()
+						var err error
+						for i := 0; i < 5; i++ {
+							ctx, cancel := context.WithDeadline(ctx, time.Now().Add(timeout))
+							upCh := make(chan *ttnpb.GatewayUp)
+							downCh := make(chan *ttnpb.GatewayDown)
+							err = ptc.Link(ctx, t, ctc.ID, ctc.Key, upCh, downCh)
+							cancel()
+							if err.Error() != "error:unknown:unknown (context deadline exceeded)" {
```

Comment:
Nah, use proper error checking, like `errors.IsDeadlineExceeded(err)`, as is also being used below

Reply:
What I understand is that the tests checks what happens when linking with/without auth. In case the auth is valid, then `ptc.Link` will return once the context deadline has been exceeded. This is OK, because it verifies that the auth check has actually been checked.

`error:unknown:unknown (context deadline exceeded)` is different: It shows that the deadline exceeded before even getting to the part of validating the gateway auth. In that sense, it is wrong to consider it a proper deadline exceeded error (because at that point the test has not finished all its checks yet).

At least that's what I gather to be the original intent of the test here. And this is why the test is repeated in case the test repeats if the context expires earlier. Merely increasing the timeout is also not a solution, because it can still lead to failures.

Reply:
Indeed, if `Link()` is successful, it doesn't return and it gets canceled via timeout, and we check if that is the error that occurred.

But what happens here exactly then? Is the context deadline exceeded before it can actually test anything?

If that is the case, shouldn't we give more time to the linking until the context times out, i.e. shouldn't the `WithTimeout()` value be higher?

What error is `err:unknown:unknown (context deadline exceeded)`? That seems like a `context.DeadlineExceeded` as cause, in an unknown error of ours.

Reply:
Based on testing, the error is returned by `GetConcentratorConfig` in https://github.com/TheThingsNetwork/lorawan-stack/blob/80083f2e26de5565ab4d3e6fa2c4067113d37c5e/pkg/gatewayserver/gatewayserver_test.go#L171

> But what happens here exactly then? Is the context deadline exceeded before it can actually test anything?

Yes

> If that is the case, shouldn't we give more time to the linking until the context times out, i.e. shouldn't the WithTimeout() value be higher?

Also yes, but this will severely slow down the tests. Also, Travis can be very slow at times and will fail even with a 2*timeout. We can also do that instead, but we will keep getting CI failures, just less frequently

Reply:
So what is your proposal?

Reply:
The simple solution I can think of is retry the test if it fails because the context expired early (the current change). 

A more elaborate approach could be to have an event on authentication success and wait for that instead (so that the test is not based on waiting for an arbitrary amount of time instead)

Reply:
> The simple solution I can think of is retry the test if it fails because the context expired early (the current change)

This is not a solution. This only reducing the probability of the flaky test.

Why not simply wait longer for the link to establish, i.e. `timeout*N` for the deadline?